hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractEvictableRecordStore.java

```diff
@@ -120,53 +127,74 @@ public boolean isExpirable() {
     }
 
     /**
-     * Intended to put an upper bound to iterations. Used in evictions.
+     * Intended to put an upper bound to sampling. Used in evictions.
      *
      * @param size       of iterate-able.
      * @param percentage percentage of size.
-     * @return 100 If calculated iteration count is less than 100, otherwise returns calculated iteration count.
+     * @return 100 If calculated sample count is less than
+     * 100, otherwise returns calculated sample count.
      */
-    private int getMaxIterationCount(int size, int percentage) {
-        final int defaultMaxIterationCount = 100;
-        final float oneHundred = 100F;
-        float maxIterationCount = size * (percentage / oneHundred);
-        if (maxIterationCount <= defaultMaxIterationCount) {
-            return defaultMaxIterationCount;
+    private int getMaxSampleCount(int size, int percentage) {
+        if (size <= MIN_TOTAL_NUMBER_OF_KEYS_TO_SCAN) {
+            return size;
         }
-        return Math.round(maxIterationCount);
+
+        int numberOfKeysInPercentage = (int) (1D * size * percentage / ONE_HUNDRED_PERCENT);
+        return Math.max(MIN_TOTAL_NUMBER_OF_KEYS_TO_SCAN, numberOfKeysInPercentage);
     }
 
-    private int evictExpiredEntriesInternal(int maxIterationCount, long now, boolean backup) {
-        int evictedEntryCount = 0;
-        int checkedEntryCount = 0;
-        initExpirationIterator();
+    private int evictExpiredEntriesInternal(int maxSample, long now, boolean backup) {
+        int sampledCount = 0;
+        int evictedCount = 0;
+        Iterator<Map.Entry<Data, Record>> iterator = initExpirationIterator();
+        do {
+            sampledCount += sampleForExpiry();
+            evictedCount += evictExpiredSamples(now, backup);
+        } while (sampledCount < maxSample && iterator.hasNext());
```

Comment:
We may leave here without draining the queue thus retraining entries in the deque that
- prevent the entry objects from getting GC'd
- make the deque to resize in the next `evictExpiredEntriesInternal()` call since it's not empty and we always add a predefined number of entries rather than checking the size of the deque
- make the next call to `evictExpiredEntriesInternal()` to work with entries collected in the previous call that may not be in the `RecordStore` anymore (unless deletions call it as well in `beforeRun()`)

I think draining the deque before returning from or when entering this method could help here.

Reply:
>We may leave here without draining the queue 
How can it be? It seems to me, we always drain the queue.